crates/edr_evm/src/state/trie/account.rs

```diff
@@ -399,153 +150,167 @@ impl AccountTrie {
         value: &U256,
         default_account_fn: &dyn Fn() -> Result<AccountInfo, ErrorT>,
     ) -> Result<Option<U256>, ErrorT> {
-        let (storage_trie_db, storage_root) =
-            self.storage_trie_dbs.entry(*address).or_insert_with(|| {
-                let storage_trie_db = Arc::new(MemoryDB::new(true));
-                let storage_root = {
-                    let mut storage_trie =
-                        Trie::new(storage_trie_db.clone(), Arc::new(HasherKeccak::new()));
-                    B256::from_slice(&storage_trie.root().unwrap())
-                };
-
-                (storage_trie_db, storage_root)
-            });
+        self.mutate()
+            .insert_storage_slot(address, index, value, default_account_fn)
+    }
+
+    /// Retrieves the trie's state root.
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip(self)))]
+    pub fn state_root(&self) -> B256 {
+        self.state_trie.root()
+    }
+
+    /// Retrieves the storage root of the account at the specified address.
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip(self)))]
+    pub fn storage_root(&self, address: &Address) -> Option<B256> {
+        self.storage_tries.get(address).map(StorageTrie::root)
+    }
 
-        let old_value = {
-            let mut storage_trie = Trie::from(
-                storage_trie_db.clone(),
-                Arc::new(HasherKeccak::new()),
-                storage_root.as_slice(),
-            )
-            .expect("Invalid storage root");
+    fn mutate(&'a mut self) -> AccountTrieMutation<'a> {
```

Comment:
It might even suffice here to do:
```suggestion
    fn mutate(&mut self) -> AccountTrieMutation<'_> {
```

Reply:
Fixed